tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -7726,6 +7733,64 @@ clang::QualType HLSLExternalSource::CheckVectorConditional(
   return ResultTy;
 }
 
+// Apply type specifier sign to the given QualType.
+// Other than privmitive int type, only allow shorthand vectors and matrices to be unsigned.
+clang::ParsedType HLSLExternalSource::ApplyTypeSpecSignToParsedType(
+    _In_ clang::ParsedType &parsedType, _In_ clang::TypeSpecifierSign TSS,
+    _In_ clang::SourceLocation Loc, _In_ const clang::PrintingPolicy &policy) {
+  if (TSS == TypeSpecifierSign::TSS_unspecified) {
+    return parsedType;
+  }
+  QualType BaseType = parsedType.get();
+  ArTypeObjectKind objKind = GetTypeObjectKind(BaseType);
+  if (objKind != AR_TOBJ_VECTOR && objKind != AR_TOBJ_MATRIX) {
+    return parsedType;
+  }
+  QualType elementType = GetMatrixOrVectorElementType(BaseType);
+  // check if element type is unsigned and check if such vector exists
+  // If not create a new one, Make a QualType of the new kind
+  ArBasicKind elementKind = GetTypeElementKind(elementType);
+  HLSLScalarType scalarType = ScalarTypeForBasic(elementKind);
+  // only ints can have a signed/unsigned types
+  if (scalarType != HLSLScalarType_int) {
+    m_sema->Diag(Loc, diag::err_sema_invalid_sign_spec)
+        << g_ArBasicTypeNames[elementKind];
+    return parsedType;
+  } else {
+    // Check given TypeSpecifierSign. If unsigned, change int to uint.
+    if (TSS == TypeSpecifierSign::TSS_unsigned) {
+      scalarType = HLSLScalarType_uint;
+    }
+    // Get new vector types for a given TypeSpecifierSign.
+    if (objKind == AR_TOBJ_VECTOR) {
+      UINT colCount = GetHLSLVecSize(BaseType);
+      QualType qt = LookupVectorType(scalarType, colCount);
+      TypedefDecl *qts = m_vectorTypedefs[scalarType][colCount - 1];
+
+      if (qts == nullptr) {
+        qts = CreateVectorSpecializationShorthand(*m_context, qt, scalarType,
+                                                  colCount);
+        m_vectorTypedefs[scalarType][colCount - 1] = qts;
```

Comment:
It's probably a good time to encapsulate this lookup + create shorthand if it doesn't exist block in a method.